johnzon-core/src/main/java/org/apache/johnzon/core/Snippet.java

```diff
@@ -0,0 +1,413 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.johnzon.core;
+
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.json.JsonValue;
+import javax.json.stream.JsonGenerator;
+import javax.json.stream.JsonGeneratorFactory;
+import java.io.ByteArrayOutputStream;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import static org.apache.johnzon.core.JsonGeneratorFactoryImpl.GENERATOR_BUFFER_LENGTH;
+
+/**
+ * Constructs short snippets of serialized JSON text representations of
+ * JsonValue instances in a way that is ideal for error messages.
+ *
+ * Instances of Snippet are thread-safe, reusable and memory-safe.  Snippet
+ * serializes only enough of the json to fill the desired snippet size and
+ * is therefore safe to use regardless of the size of the JsonValue.
+ */
+public class Snippet {
+
+    private final int max;
+    private final JsonGeneratorFactory generatorFactory;
+
+    /**
+     * This constructor should be used only in static or other scenarios were
+     * there is no JsonGeneratorFactory instance in scope.
+     *
+     * @param max the maximum length of the serialized json produced via of()
+     */
+    public Snippet(final int max) {
+        this(max, new JsonGeneratorFactoryImpl(new HashMap<String, Object>() {
+            {
+                this.put(GENERATOR_BUFFER_LENGTH, max);
+            }
+        }));
+    }
+
+    /**
+     * This is the preferred approach to using Snippet in any context where
+     * there is an existing JsonGeneratorFactory in scope.
+     *
+     * @param max the maximum length of the serialized json produced via of()
+     * @param generatorFactory the JsonGeneratorFactory created by the user
+     */
+    public Snippet(final int max, final JsonGeneratorFactory generatorFactory) {
+        this.max = max;
+        this.generatorFactory = generatorFactory;
+    }
+
+    /**
+     * Create a serialized json representation of the supplied
+     * JsonValue, truncating the value to the specified max length.
+     * Truncated text appears with a suffix of "..."
+     *
+     * This method is thread safe.
+     *
+     * @param value the JsonValue to be serialized as json text
+     * @return a potentially truncated json text
+     */
+    public String of(final JsonValue value) {
+        try (final Buffer buffer = new Buffer()) {
+            buffer.write(value);
+            return buffer.get();
```

Comment:
was expecting that moving get() after the auto-close try block would auto-flush it, isn't it right?

Reply:
I prefer to have the buffer instance constructed inside the try-with-resources as I like the assurance there can be 100% certain there can be no possible leak.  If we don't like flush, we could call close which is documented to idempotent and safe to call more than once.  Johnzon properly implements idempotent closes so it is safe.  If a JSON-P impl doesn't properly implement close and fails, that'd be their bug IMO.

That said, I'm ok if you'd like to adjust it.